File: mjlib_c++/hulib.cpp
#include <stdio.h>
#include <string.h>
#include "hulib.h"
#include "table_mgr.h"

#define LOG printf
//#define LOG log

void log(char*, ...)
{
}

bool HuLib::get_hu_info(char* const hand_cards, Wave* const waves, char self_card_index, char other_card_index, int gui_index)
{
    char hand_cards_tmp[34];
    memcpy(hand_cards_tmp, hand_cards, 34);

    if(self_card_index != 34)
    {
        hand_cards_tmp[self_card_index]++;
    }
    else if(other_card_index != 34)
    {

        hand_cards_tmp[other_card_index]++;
    }

    int gui_num = hand_cards_tmp[gui_index];
    hand_cards_tmp[gui_index] = 0;

    ProbabilityItemTable ptbl;
    if(!split(hand_cards_tmp, gui_num, ptbl))
    {
        return false;
    }

    return check_probability(ptbl, gui_num);
}

bool HuLib::split(char* const cards, int gui_num, ProbabilityItemTable& ptbl)
{
    if(!_split(cards, gui_num, 0,  0,  8, true,  ptbl)) return false;
    if(!_split(cards, gui_num, 1,  9, 17, true,  ptbl)) return false;
    if(!_split(cards, gui_num, 2, 18, 26, true,  ptbl)) return false;
    if(!_split(cards, gui_num, 3, 27, 33, false, ptbl)) return false;

    return true;
}

bool HuLib::_split(char* const cards, int gui_num, int color, int min, int max, bool chi, ProbabilityItemTable& ptbl)
{
    int key = 0;
    int num = 0;

    for(int i = min; i <= max; ++i)
    {
        key = key*10 + cards[i];
        num = num + cards[i];
    }


    if(num > 0)
    {
        if(!list_probability(color, gui_num, num, key, chi, ptbl))
        {
            return false;
        }
    }

    return true;
}

bool HuLib::list_probability(int color, int gui_num, int num, int key, bool chi, ProbabilityItemTable& ptbl)
{
    int anum = ptbl.array_num;
    for(int i=0; i<=gui_num; ++i)
    {
        int yu = (num + i)%3;
        if(yu == 1) continue;
        bool eye = (yu == 2);
        if(TableMgr::get_instance()->check(key, i, eye, chi))
        {
            ProbabilityItem& item = ptbl.m[anum][ptbl.m_num[anum]];
            ptbl.m_num[anum]++;

            item.eye = eye;
            item.gui_num = i;
        }
    }

    LOG("gui_num = %d, color = %d key = %d num = %d pra_num=%d\n", gui_num, color, key, num, ptbl.m_num[anum]);

    if(ptbl.m_num[anum] <= 0)
    {
        return false;
    }

    ptbl.array_num++;
    return true;
}

bool HuLib::check_probability(ProbabilityItemTable& ptbl, int gui_num)
{
    LOG("组合 array_num = %d\n", ptbl.array_num);
    // 全是鬼牌
    if(ptbl.array_num == 0)
    {
        return gui_num >= 2;
    }

    // 只有一种花色的牌的鬼牌
    if(ptbl.array_num == 1) return true;

    // 尝试组合花色，能组合则胡
    for(int i=0; i<ptbl.m_num[0]; ++i)
    {
        ProbabilityItem& item = ptbl.m[0][i];
        bool eye = item.eye;
        int gui = gui_num - item.gui_num;
        if(check_probability_sub(ptbl, eye, gui, 1, ptbl.array_num))
        {
            return true;
        }
    }
    return false;
}

bool HuLib::check_probability_sub(ProbabilityItemTable& ptbl, bool& eye, int& gui_num, int level, int max_level)
{
    for(int i=0; i<ptbl.m_num[level]; ++i)
    {
        ProbabilityItem& item = ptbl.m[level][i];

        if(eye && item.eye) continue;

        if(gui_num < item.gui_num) continue;

        if(level < max_level - 1)
        {
            int old_gui_num = gui_num;
            bool old_eye = eye;
            gui_num -= item.gui_num;
            eye = eye || item.eye;

            if(check_probability_sub(ptbl, eye, gui_num, level + 1, ptbl.array_num))
            {
                return true;
            }

            eye = old_eye;
            gui_num = old_gui_num;
            continue;
        }

        if(!eye && !item.eye && item.gui_num + 2 > gui_num) continue;
        return true;
    }

    return false;
}

bool HuLib::check_7dui(char* const cards)
{
    int c = 0;
    for(int i=0;i<34;++i)
    {
        if(cards[i] % 2 != 0) return false;
        c += cards[i];
    }

    if(c != 34) return false;

    return true;
}

bool HuLib::check_pengpeng(char* const cards, Wave* const waves, char self_card, char other_card)
{
    bool eye = false;

    for(int i=0;i<34;++i)
    {
        int c = cards[i];
        if(c == 1 || c == 4) return false;

        if(c == 2)
        {
            if(eye) return false;
            eye = true;
        }
    }

    // 检查吃碰杠的牌

    return true;
}

File: mjlib_lua/gui/hulib.lua
local mjlib = require "mjlib"
local table_mgr = require "table_mgr"

local split_table = {
    {min = 1,  max = 9,  chi = true},
    {min = 10, max = 18, chi = true},
    {min = 19, max = 27, chi = true},
    {min = 28, max = 34, chi = false}
}

local M = {}
function M.check_7dui(hand_cards, waves)
    if #waves > 0 then return false end

    for _,c in ipairs(hand_cards) do
        if c % 2 ~= 0 then
            return false
        end
    end

    return true
end

function M.check_pengpeng()

end

function M.get_hu_info(cards, waves, gui_index)
    local hand_cards = {}
    for i,v in ipairs(cards) do
        hand_cards[i] = v
    end

    local gui_num = 0
    if gui_index > 0 then
        gui_num = hand_cards[gui_index]
        hand_cards[gui_index] = 0
    end

    local splited_tbl = M.split_info(hand_cards, gui_num)
    if not splited_tbl then
        return false
    end

    return M.check_probability(splited_tbl, gui_num)
end

function M.list_probability(gui_num, num, key, chi)
    local find = false
    local t = {}
    for i=0, gui_num do
        local yu = (num + i)%3
        if yu ~= 1 then
            local eye = (yu == 2)
            if find or table_mgr:check(key, i, eye, chi) then
                table.insert(t, {eye = eye, gui_num = i})
                find = true
            end
        end
    end

    return t
end

-- 根据花色切分
function M.split_info(t, gui_num)
    local ret = {}
    for _,v in ipairs(split_table) do
        local key = 0
        local num = 0
        for i=v.min,v.max do
            key = key*10 + t[i]
            num = num + t[i]
        end
        if num > 0 then
            local t = M.list_probability(gui_num, num, key, v.chi)
            if #t == 0 then
                return false
            end
            table.insert(ret, t)
        end
    end
    return ret
end

function M.check_probability_sub(splited_table, eye, gui_num, level, max_level)
    for _,v in ipairs(splited_table[level]) do
        repeat
            if eye and v.eye then
                break
            end

            if gui_num < v.gui_num then
                break
            end

            if level < max_level then
                if M.check_probability_sub(
                    splited_table,
                    eye or v.eye,
                    gui_num - v.gui_num,
                    level + 1,
                    max_level) then
                    return true
                end
                break
            end

            if not eye and not v.eye and gui_num < 2 + v.gui_num then
                break
            end

            return true
       until(true)
    end

    return false
end

function M.check_probability(splited_table, gui_num)
    local c = #splited_table
    -- 全是鬼牌
    if c == 0 then
        return true
    end

    -- 只有一种花色的牌和鬼牌
    if c == 1 then
        return true
    end

    -- 组合花色间的组合，如果能满足组合条件，则胡
    for i,v in ipairs(splited_table[1]) do
        local ret = M.check_probability_sub(splited_table, v.eye, gui_num - v.gui_num, 2, c)
        if ret then
            return true
        end
    end

    return false
end

return M
